### PR TITLE
Add bindings for git_merge()

### PR DESCRIPTION
**Pull request by tksfz**

### CODE REVIEW

**Comment by tksfz** [comment]
Hi I think this is at least ready to get some feedback.  I'm only trying to build enough to get git_merge() implemented which requires AnnotatedCommit and MergeOptions.  There are several other merge-related functions, as well as cherry-pick and rebase functions.  I'm hoping they'll get easier to implement once the basics are laid down.

Feel free to give any kind of feedback or let me know if this is not the right direction.

File: src/lib.rs [hunk 1]
```diff
@@ -230,6 +231,19 @@ pub enum ConfigLevel {
     Highest,
 }
 
+/// An enumeration for favoring sides to resolve merge conflicts
+#[derive(PartialEq, Eq, Debug, Copy, Clone)]
+pub enum FileFavor {
+    /// Record conflicts in the index without favoring sides
```
Comment (alexcrichton) [inline src/lib.rs]: Could these docs copy over the longer form docs in the corresponding header file?

File: src/merge.rs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+use libc::c_uint;
+
+use {raw, Object, FileFavor};
+use util::Binding;
+use call::Convert;
+
+/// annotated commits, the input to merge and rebase
+pub struct AnnotatedCommit<'repo> {
+    raw: *mut raw::git_annotated_commit,
+    _marker: marker::PhantomData<Object<'repo>>,
```
Comment (alexcrichton) [inline src/merge.rs]: Looking at the definition in C, it looks like this actually contains a `Commit`, so could this phantom data be `Commit<'repo>` instead?

File: src/merge.rs [hunk 3]
```diff
@@ -0,0 +1,138 @@
+use std::marker;
+use std::mem;
+use libc::c_uint;
+
+use {raw, Object, FileFavor};
+use util::Binding;
+use call::Convert;
+
+/// annotated commits, the input to merge and rebase
```
Comment (alexcrichton) [inline src/merge.rs]: Could this be expanded a bit (and use the same sentence-like style of other docs)?

File: src/merge.rs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    /// Take extra time to find minimal diff
+    pub fn minimal(&mut self, find: bool) -> &mut MergeOptions {
+        self.flag(raw::GIT_MERGE_FILE_DIFF_MINIMAL, find)
+    }
+
+    /// Acquire a pointer to the underlying raw options.
+    ///
+    /// This function is unsafe as the pointer is only valid so long as this
+    /// structure is not moved, modified, or used elsewhere.
+    // modeled after DiffOptions.raw()
```
Comment (alexcrichton) [inline src/merge.rs]: It's ok to drop these kinds of comments

File: src/merge.rs (hunk 2 above)
Comment (alexcrichton) [inline src/merge.rs]: Also, could some accessor functions of `AnnotatedCommit` also be exposed? It looks like there's a few simple-ish ones in the header files

File: src/repo.rs [hunk 5]
```diff
@@ -785,6 +786,17 @@ impl Repository {
         }
     }
 
+    /// Creates a git_annotated_commit from the given reference.
+    pub fn reference_to_annotated_commit(&self, reference: &Reference) -> Result<AnnotatedCommit, Error> {
```
Comment (alexcrichton) [inline src/repo.rs]: Could you make sure this wraps at 80 chars?

File: src/repo.rs [hunk 6]
```diff
@@ -989,6 +1001,36 @@ impl Repository {
         Ok(())
     }
 
+    /// Merges the given commit(s) into HEAD, writing the results into the working directory.
+    /// Any changes are staged for commit and any conflicts are written to the index. Callers
+    /// should inspect the repository's index after this completes, resolve any conflicts and
+    /// prepare a commit.
```
Comment (alexcrichton) [inline src/repo.rs]: Could this included the rest of the docs as well? (e.g. just copied from libgit2), also, can this be wrapped to 80 chars?

File: src/repo.rs [hunk 7]
```diff
@@ -989,6 +1001,36 @@ impl Repository {
...
+    pub fn merge(&self,
+                 annotated_commits: &[&AnnotatedCommit],
+                 merge_opts: &MergeOptions,
```
Comment (alexcrichton) [inline src/repo.rs]: I think that this can technically be `Option<&MergeOptions>`

File: src/repo.rs [hunk 8]
```diff
@@ -989,6 +1001,36 @@ impl Repository {
...
+            try_call!(raw::git_checkout_init_options(&mut raw_checkout_opts,
+                                raw::GIT_CHECKOUT_OPTIONS_VERSION));
+            match checkout_opts {
+                Some(c) => c.configure(&mut raw_checkout_opts),
+                None => {}
+            }
+
+            let commit_ptrs: Vec<*const raw::git_annotated_commit> =  annotated_commits.iter().map(|c| {
+                c.raw() as *const raw::git_annotated_commit
+            }).collect();
```
Comment (alexcrichton) [inline src/repo.rs]: Stylistically I'd recommend putting the type annotation at the end and doing something like `.collect::<Vec<_>>()`

**Comment by alexcrichton** [comment]
Thanks @tksfz!

File: src/lib.rs (hunk 1 above)
Comment (tksfz) [inline src/lib.rs]: Yes I'll change it to copy the docs from libgit2.  I was under the impression that other docs in this code base seemed to be paraphrasing the libgit2 docs but I like copying them.

File: src/merge.rs (hunk 2 above)
Comment (tksfz) [inline src/merge.rs]: re phantom data: interesting, i'll take a look.  I vaguely recall that you can create annotated commits from things other than commits but I may well be mistaken.

File: src/merge.rs (hunk 4 above)
Comment (tksfz) [inline src/merge.rs]: Ok, I noticed other raw() methods were doing the same thing.  I'll remove all of it except for the first line.

File: src/merge.rs (hunk 3 above)
Comment (tksfz) [inline src/merge.rs]: done

File: src/merge.rs (hunk 4 above)
Comment (tksfz) [inline src/merge.rs]: done

File: src/repo.rs (hunk 6 above)
Comment (tksfz) [inline src/repo.rs]: done

**Comment by tksfz** [comment]
Re accessor functions for annotated_commit - are you referring to the functions that create a new annotated_commit (get_annotated_commit_from_{fetchhead, revspec}, etc.) or the get_annotated_commit_id() function?

**Comment by tksfz** [comment]
Re `Option<&MergeOptions>` - can you clarify?  How would I generate a raw_merge_options for that, in a clean way?  I think I'd need to call `git_merge_init_options` at least.  Currently, I'm supposing that callers will call `MergeOptions::new()` to create it.

File: src/merge.rs (hunk 2 above)
Comment (tksfz) [inline src/merge.rs]: I changed it to `Commit<`repo>`

File: src/repo.rs (hunk 6 above)
Comment (tksfz) [inline src/repo.rs]: Done - it's kind of long now though.

**Comment by alexcrichton** [comment]
> Re accessor functions for annotated_commit - are you referring to the functions that create a new annotated_commit (get_annotated_commit_from_{fetchhead, revspec}, etc.) or the get_annotated_commit_id() function?

Ah I mostly meant the accessors, but I do think that some extra constructors would be nice as well (just so the methods of creating them are all bound).

>  How would I generate a raw_merge_options for that, in a clean way?

In the `None` case I believe you can just pass down a null pointer.

**Comment by tksfz** [comment]
Re None for MergeOptions - ah ok great passing null seems to work great in that case.

Re accessors:  Ok, I added the only accessor which is for id().

I haven't added the constructors yet but I'll likely have a branch in the future to add that.

Btw, I already started another follow-on branch that adds Repository::cleanup_state(), which is kind of necessary for using merge, as well as a few other functions.  For the sake of keeping this branch small I figured I'd keep this branch as-is, and create another pull for that branch.

**Comment by alexcrichton** [comment]
Thanks!

**Pull request merged**